### PR TITLE
feat(project): add new type of lists (recommendations)

### PR DESCRIPTION
**Pull request by AntonLantukh**
## Description

1. Add new "content_list" type support (recommendations).
2. Support it for Landing Page and Menu.

### Steps completed:

According to our definition of done, I have completed the following steps:

- [ ] Acceptance criteria met
- [ ] Unit tests added
- [ ] Docs updated (including config and env variables)
- [ ] Translations added
- [ ] UX tested
- [ ] Browsers / platforms tested
- [ ] Rebased & ready to merge without conflicts
- [ ] Reviewed own code

### CODE REVIEW

**Comment by AntonLantukh** [comment]
App Config for testing: `0i9cflkq`

File: packages/ui-react/src/pages/ScreenRouting/RecommendationsScreenRouter.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import ErrorPage from '../../components/ErrorPage/ErrorPage';
+import type { ScreenComponent } from '../../../types/screens';
+
+import PlaylistGrid from './sharedScreens/PlaylistGrid/PlaylistGrid';
+
+export const playlistScreenMap = new ScreenMap<Playlist, ScreenComponent<Playlist>>();
+
+playlistScreenMap.registerDefault(PlaylistGrid);
+
+const RecommendationsScreenRouter = () => {
```
Comment (ChristiaanScheermeijer) [inline packages/ui-react/src/pages/ScreenRouting/RecommendationsScreenRouter.tsx]: Do we need a separate router or could we re-use `<PlaylistScreenRouter />` and pass the list type along from a prop?

**Review by ChristiaanScheermeijer — COMMENTED** [review]
The PR overall looks good, but I do have some questions:

- ~Do you have content list IDs to test with?~ You just added it 😄 
- Do we need to add e2e testing for this?
- Do we really need an abstraction for the content and playlist APIs or could it just be two separate services?
  - This could prevent some breaking changes in the `usePlaylist` hook for example
- Should we use different names for the service/controller/types?
  - Content list can be confused with contentTypes
  - `RecommendationsService` sounds like it's used for different purposes
  - I suggest renaming it to `PlaylistController`, `PlaylistService`, `PlaylistContentService` and `PLAYLIST_TYPE` removing the content analogy from most most places

File: packages/ui-react/src/pages/ScreenRouting/RecommendationsScreenRouter.tsx (hunk 1 above)
Comment (AntonLantukh) [reply packages/ui-react/src/pages/ScreenRouting/RecommendationsScreenRouter.tsx]: So for menu items you think we could pass 'type' as a query param? We could reuse it. I was just thinking that it could be cleaner when we use a separate router. We can also tweak translations and avoid using LiveChannels screen.

**Comment by AntonLantukh** [comment]
@ChristiaanScheermeijer 

 > Do we need to add e2e testing for this?

We do need. I will add it to the Q3 plan. 

> Do we really need an abstraction for the content and playlist APIs or could it just be two separate services? 
> This could prevent some breaking changes in the `usePlaylist` hook for example.

Do you mean having two hooks "usePlaylist" and "usePlaylistContent" which could be called conditionally + removing the controller completely? We still need to know the type to make a decision, so there will be also a "parent" hook receiving it as an argument.

> Should we use different names for the service/controller/types?

I agree naming is a problem. My idea was to consume playlists, content lists (recommendations), series using controller in the future. That is why I used "ContentController" name. We also have some plans to deliever an additional "media" endpoint returning data based on the content type (maybe series, stream data inside). Though it can be also overengineering to consider this case in advance and to avoid braking changes we could start with conditional calls.

File: packages/common/src/services/content/RecommendationsService.ts
Comment (budzworthy) [inline packages/common/src/services/content/RecommendationsService.ts:1]: Do we need to differentiate content lists from recommendations lists?

File: packages/common/src/services/content/ContentService.ts
Comment (budzworthy) [inline packages/common/src/services/content/ContentService.ts:1]: I think we should avoid putting the Media vs. Content abstraction into the web app for now. It adds an extra layer above playlist and content lists that may make the sample card harder to understand.

File: packages/common/src/services/content/PlaylistService.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    const response = await fetch(url);
+    const data = (await getDataOrThrow(response)) as Playlist;
+
+    return this.transformPlaylist(data, params.related_media_id);
+  };
+
+  /**
+   * Get playlist by id
+   */
+  getContentList = async ({ id, params }: { id: string | undefined; params: GetContentListParams }): Promise<Playlist | undefined> => {
```
Comment (budzworthy) [inline packages/common/src/services/content/PlaylistService.ts]: I'm not sure this boilerplate is adding value. I would prefer to keep the getPlaylist and any other new api endpoints in the ApiService.

**Review by budzworthy — CHANGES_REQUESTED** [review]
I'm still going through some parts, but I think we can keep this change simpler without adding the content service abstractions. If we add the new content list endpoint to the existing Api Service, we can do switching at the hook or controller level to get playlists or content lists as needed.

File: packages/common/src/services/content/ContentService.ts
Comment (AntonLantukh) [reply packages/common/src/services/content/ContentService.ts:1]: I personally like that we define the contract in the Service and let the controller select the right endpoint. It adds complexity but looks clean. Though I also agree that maybe it is still early for more global refactoring here.

**Comment by AntonLantukh** [comment]
@ChristiaanScheermeijer @dbudzins simplified approach without abstraction layer.

File: packages/common/src/services/ApiService.ts [hunk 3]
```diff
@@ -44,20 +45,64 @@ export default class ApiService {
...
+  };
+
+  /**
+   * Transform incoming playlists
+   */
+  private transformPlaylist = (playlist: Playlist, relatedMediaId?: string) => {
+    playlist.playlist = playlist.playlist.map((item) => this.transformMediaItem(item, playlist));
+
+    // remove the related media item (when this is a recommendations playlist)
+    if (relatedMediaId) playlist.playlist.filter((item) => item.mediaid !== relatedMediaId);
```
Comment (budzworthy) [inline packages/common/src/services/ApiService.ts]: https://www.imperialviolet.org/2014/02/22/applebug.html
Also, I think you need to assign for filter to have an effect.
```suggestion
    if (relatedMediaId) {
      playlist.playlist = playlist.playlist.filter((item) => item.mediaid !== relatedMediaId);
    }
```

File: packages/common/types/config.ts [hunk 4]
```diff
@@ -39,20 +41,20 @@ export type Drm = {
...
+  type: PlaylistType;
   featured?: boolean;
   backgroundColor?: string | null;
 };
 
 export type Menu = {
   label: string;
   contentId: string;
-  type?: Extract<ContentType, 'playlist'>;
+  type?: Extract<PlaylistType, 'playlist' | 'content_list'>;
```
Comment (budzworthy) [inline packages/common/types/config.ts]: This can be easier since you have a type for it now:
```suggestion
  type?: PLAYLIST_TYPE.playlist | PLAYLIST_TYPE.content_list;
```

File: packages/common/types/playlist.ts [hunk 5]
```diff
@@ -27,41 +25,15 @@ export type PlaylistItem = {
...
-  scheduleDemo?: string;
-  catchupHours?: string;
-  mediaStatus?: MediaStatus;
-  scheduledStart?: Date;
-  scheduledEnd?: Date;
-  markdown?: string;
-  scheduleType?: string;
-  [key: string]: unknown;
-};
+} & CustomParams;
```
Comment (budzworthy) [inline packages/common/types/playlist.ts:36]: Nice!

File: packages/common/src/utils/urlFormatting.ts [hunk 6]
```diff
@@ -103,8 +103,8 @@ export const mediaURL = ({
   return createPath(PATH_MEDIA, { id: media.mediaid, title: slugify(media.title) }, { r: playlistId, play: play ? '1' : null, e: episodeId });
 };
 
-export const playlistURL = (id: string, title?: string) => {
-  return createPath(PATH_PLAYLIST, { id, title: title ? slugify(title) : undefined });
+export const playlistURL = ({ id, title, type }: { id: string; title?: string; type?: 'playlist' | 'content_list' }) => {
+  return createPath(PATH_PLAYLIST, { id, title: title ? slugify(title) : undefined }, { t: type });
```
Comment (budzworthy) [inline packages/common/src/utils/urlFormatting.ts]: I would keep this as succinct as possible and try to avoid changing the existing playlist routing:
```suggestion
  return createPath(PATH_PLAYLIST, { id, title: title ? slugify(title) : undefined }, type === 'content_list' ? { t: 'c' } : {});
```

File: packages/hooks-react/src/usePlaylists.ts
Comment (budzworthy) [inline packages/hooks-react/src/usePlaylists.ts:1]: We should probably only have usePlaylist (singular) or usePlaylists (plural) otherwise there's a big risk of code divergence (seen here with different query keys whether content lists are retrieved with this hook or the useContentList hook)

File: packages/hooks-react/src/usePlaylistContent.ts
Comment (budzworthy) [inline packages/hooks-react/src/usePlaylistContent.ts:1]: I think the approach you took in usePlaylists was probably more maintainable. 1 hook that has a tiny bit of playlist and content list logic is better than 3 hooks.

File: packages/ui-react/src/pages/ScreenRouting/sharedScreens/PlaylistGrid/PlaylistGrid.tsx
Comment (budzworthy) [inline packages/ui-react/src/pages/ScreenRouting/sharedScreens/PlaylistGrid/PlaylistGrid.tsx:1]: I would keep the 2 concepts of media and lists here, instead of adding shared. Shared can be confusing because it's unclear if it's shared between media and playlist, media and content list, or playlist and content list. If you want to rename playlistScreens to listScreens that's fine as a compromise.

File: packages/common/src/services/ApiService.ts (hunk 3 above)
Comment (AntonLantukh) [reply packages/common/src/services/ApiService.ts]: Hm, looks like this is the way it worked before. I will assign it.

File: packages/ui-react/src/pages/ScreenRouting/sharedScreens/PlaylistGrid/PlaylistGrid.tsx
Comment (AntonLantukh) [reply packages/ui-react/src/pages/ScreenRouting/sharedScreens/PlaylistGrid/PlaylistGrid.tsx:1]: I think this is what is left after removing "RecsScreen". I will move it to Playlists.

File: packages/common/src/services/ApiService.ts (hunk 3 above)
Comment (budzworthy) [reply packages/common/src/services/ApiService.ts]: Probably just a bug then.. good to fix while we're here

File: packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx [hunk 7]
```diff
@@ -37,6 +36,12 @@ const PlaylistScreenRouter = () => {
     return <ErrorPage title={t('empty_shelves_heading')} message={t('empty_shelves_description')} />;
   }
 
+  // register playlist screens
+  playlistScreenMap.registerDefault(PlaylistGrid);
+  if (type === PLAYLIST_TYPE.playlist) {
+    playlistScreenMap.registerByContentType(PlaylistLiveChannels, PLAYLIST_CONTENT_TYPE.live);
+  }
```
Comment (ChristiaanScheermeijer) [inline packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx]: This feels a bit hacky, why is this needed?

File: packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx [hunk 8]
```diff
@@ -2,27 +2,26 @@ import React from 'react';
...
-// register playlist screens
-playlistScreenMap.registerByContentType(PlaylistLiveChannels, CONTENT_TYPE.live);
-playlistScreenMap.registerDefault(PlaylistGrid);
-
 const PlaylistScreenRouter = () => {
   const params = useParams();
   const id = params.id || '';
-  const { isLoading, isFetching, error, data } = usePlaylist(id);
+  const type = useQueryParam('t') === 'c' ? PLAYLIST_TYPE.content_list : PLAYLIST_TYPE.playlist;
+
```
Comment (ChristiaanScheermeijer) [inline packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx]: Hmm, I have doubts about using query params for this. Do we include this in the canonical URL?

Alternatively, we could do this:

```tsx
export const PATH_PLAYLIST_CONTENT = '/p/:id/c/*' // or something similar

// ...

<Route path={PATH_PLAYLIST_CONTENT} element={<PlaylistScreenRouter type="content" />} />
<Route path={PATH_PLAYLIST} element={<PlaylistScreenRouter type="playlist" />} />
```

File: packages/hooks-react/src/usePlaylist.ts [hunk 9]
```diff
@@ -1,42 +1,84 @@
...
+  type,
+  params = {},
+  enabled = true,
+  usePlaceholderData = true,
+}: {
+  contentId?: string;
+  type: PlaylistMenuType;
+  params?: GetPlaylistParams;
+  enabled?: boolean;
+  usePlaceholderData?: boolean;
```
Comment (ChristiaanScheermeijer) [inline packages/hooks-react/src/usePlaylist.ts]: Wouldn't it make more sense to leave the `usePlaylist` as is and introduce a `useContentList` hook?

Comment (ChristiaanScheermeijer) [reply packages/hooks-react/src/usePlaylist.ts]: I'm now seeing @dbudzins feedback. Although I agree to that, this change will result in many changes and could be considered a breaking change. I would very much like to prevent that, but it's not a blocking issue.

File: packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx (hunk 8 above)
Comment (AntonLantukh) [reply packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx]: I think we would prefer to avoid changing playlist routing, so I was choosing between a separate screen routing and a query param. If we change the path, then I think we should also change it for the playlist.

File: packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx (hunk 7 above)
Comment (AntonLantukh) [reply packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx]: Not to register the screen for content_list where live screen is not supported yet. For now we define the screen based on the item itself, but the item doesn't tell anything about the type (though maybe we could add additional "type" property for both content list and playlists).

File: packages/hooks-react/src/usePlaylist.ts (hunk 9 above)
Comment (AntonLantukh) [reply packages/hooks-react/src/usePlaylist.ts]: That was the initial solution here:

https://github.com/jwplayer/ott-web-app/pull/556/commits/39c9f981fc88c9f04169018167224415c949102a

Would you prefer this one?

@dbudzins Danny suggested to use "usePlaylist" directly without additional hook. Not sure I have any strong preference here. I used hook approach with search before, worked fine.

Comment (AntonLantukh) [reply packages/hooks-react/src/usePlaylist.ts]: Breaking change is a valid point here. Maybe we can start with hooks here and refactor usePlaylist / usePlaylists later.

File: packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx (hunk 8 above)
Comment (budzworthy) [reply packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx]: Probably safest to keep them as separate paths. Why not just `/p/:id/*` and `/n/:id/*` (using n for coNtent list, because we might want to reserve c for individual content items if we ever get to that refactor.) I worry `/p/:id/c/*` will have collisions with `/p/:id/*` if we aren't careful.

File: packages/hooks-react/src/usePlaylist.ts (hunk 9 above)
Comment (budzworthy) [reply packages/hooks-react/src/usePlaylist.ts]: What is the breaking change? If it's just the addition of the type param, can we default it to `playlist` if not provided?

Comment (ChristiaanScheermeijer) [reply packages/hooks-react/src/usePlaylist.ts]: We consume these hooks in all platforms. The breaking change is that the hook signature changes. It would be safe if it was:

```ts
export default function usePlaylist(playlistId?: string, params: GetPlaylistParams = {}, enabled: boolean = true, usePlaceholderData: boolean = true, type: PlaylistMenuType = PLAYLIST_TYPE.playlist) {
```

Although this require adding all optional arguments with the default values when only changing the type.

It's not a huge deal, but something to be aware of.

File: packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx (hunk 8 above)
Comment (ChristiaanScheermeijer) [reply packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx]: That works for me 😄

File: packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx (hunk 7 above)
Comment (AntonLantukh) [reply packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx]: It will work then this way (out of component's scope):

```
// register playlist screens
playlistScreenMap.registerDefault(PlaylistGrid);
playlistScreenMap.register(PlaylistLiveChannels, (item) => item?.contentType === PLAYLIST_CONTENT_TYPE.live && item.type === PLAYLIST_TYPE.playlist);

```

Comment (ChristiaanScheermeijer) [reply packages/ui-react/src/pages/ScreenRouting/PlaylistScreenRouter.tsx]: We don't need a separate router screen per se, but perhaps we can create a separate screen map for content? 

```ts
const screenMap = type === 'playlist' ? playlistScreenMap : contentScreenMap;
```

The `contentScreenMap` only contains the default screen using the PlaylistGrid.

**Comment by AntonLantukh** [comment]
@ChristiaanScheermeijer @dbudzins 

1. Switched to a separate endpoint for content lists.
2. Reused PlaylistScreen for content lists.
3. Changed interface of `usePlaylist` and added optional params (was needed in one place).
4. Separated query options into a separate function to reuse in `usePlaylists` / `usePlaylist`

**Review by ChristiaanScheermeijer — APPROVED** [review]
Perfect @AntonLantukh! This looks good to me.

Only the `playlistURL` method has changed, but that's only used once in our platforms 😄

**Comment by AntonLantukh** [comment]
@ChristiaanScheermeijer rolled back `playlistURL` to the previous interface.

File: packages/hooks-react/src/usePlaylist.ts [hunk 10]
```diff
@@ -1,42 +1,78 @@
...
+  enabled: boolean = true,
+  usePlaceholderData: boolean = true,
+  type: PlaylistMenuType = PLAYLIST_TYPE.playlist,
+) {
+  const queryClient = useQueryClient();
+  const siteId = useConfigStore((state) => state.config.siteId);
+
+  const queryOptions = getPlaylistQueryOptions({ type, contentId, siteId, params, queryClient, enabled, usePlaceholderData });
+
+  return useQuery<Playlist | undefined, ApiError>(queryOptions);
```
Comment (budzworthy) [inline packages/hooks-react/src/usePlaylist.ts:77]: You've gotta add usePlacehoderData and params inputs to usePlaylists, but then you avoid the getQueryOptions construct, which is a bit non-standard I think.

(We can do this later. I know you want to get the changes out, and the double playlist hooks wasn't added from this PR)

```suggestion
  return usePlaylists([{contentId, type}], enabled ? 1 : 0, usePlaceholderData, params)[0];
```

File: platforms/web/src/containers/AppRoutes/AppRoutes.tsx [hunk 11]
```diff
@@ -24,6 +24,7 @@ export default function AppRoutes() {
         <Route element={<Layout />} errorElement={<RootErrorPage />}>
           <Route index element={<Home />} />
           <Route path={PATH_PLAYLIST} element={<PlaylistScreenRouter />} />
+          <Route path={PATH_CONTENT_LIST} element={<PlaylistScreenRouter />} />
```
Comment (budzworthy) [inline platforms/web/src/containers/AppRoutes/AppRoutes.tsx]: I would suggest we either use @ChristiaanScheermeijer 's suggestion to have the `type` as a param on PlaylistScreenRouter or alternatively, use a regex for the path and share one <Route> for playlist and content list. I think the former is preferred, because you don't need the extra path match checking inside PlaylistScreenRouter.

**Review by budzworthy — APPROVED** [review]
2 small comments. The one about the Routes and PlaylistScreenRouter component, I'd strongly suggest doing, but I'll approve as-is, because I know you're pressed for time.

File: packages/hooks-react/src/usePlaylist.ts (hunk 10 above)
Comment (AntonLantukh) [reply packages/hooks-react/src/usePlaylist.ts:77]: I didn't want to do it because of the different use case we have here. usePlaylists is intended to work with a list of shelves on the Landing Page. So it also covers the cases of Favorites and Continue Watching, while usePlaylist only supports `content_list` and `playlist` types which makes the purpose a bit different.

**Review by budzworthy — APPROVED** [review]
lgtm
